Inferences/BackwardDemodulation.cpp

```diff
@@ -168,32 +169,41 @@ struct BackwardDemodulation::ResultFn
     }
 
     if(_parent.getOptions().demodulationRedundancyCheck() && qr.literal->isEquality() &&
-      (qr.term==*qr.literal->nthArgument(0) || qr.term==*qr.literal->nthArgument(1)) ) {
+      (qr.term==*qr.literal->nthArgument(0) || qr.term==*qr.literal->nthArgument(1)) && 
+      // encompassment has issues only with positive units
+      (!_encompassing || (qr.literal->isPositive() && qr.clause->length() == 1))) {
       TermList other=EqHelper::getOtherEqualitySide(qr.literal, qr.term);
       Ordering::Result tord=_ordering.compare(rhsS, other);
       if(tord!=Ordering::LESS && tord!=Ordering::LESS_EQ) {
-        TermList eqSort = SortHelper::getEqualityArgumentSort(qr.literal);
-        Literal* eqLitS=Literal::createEquality(true, lhsS, rhsS, eqSort);
-        bool isMax=true;
-        Clause::Iterator cit(*qr.clause);
-        while(cit.hasNext()) {
-          Literal* lit2=cit.next();
-          if(qr.literal==lit2) {
-            continue;
+        if (_encompassing) {
+          if (qr.substitution->isRenamingOn(lhs,false /* we talk of a non-result, i.e., a query term */)) {
+            // under _encompassing, we know there are no other literals in qr.clause
```

Comment:
This doesn't capture the full generality of the rule as given in the paper. The substitution can be a renaming so long as we are rewriting a proper subterm (or the smaller side of the equality). Perhaps we don't want / need the full generality?

Reply:
The same reason is in Forward?

Reply:
Your comment there clarified. Think it is fine now.